transport/tracer_test.go

```diff
@@ -0,0 +1,301 @@
+		},
+		Tracer: tracer,
+	})
+
+	json.Register(rpc, json.Procedure("echo", echo))
+
+	client := json.New(rpc.Channel("yarpc-test"))
+
+	rpc.Start()
+	defer rpc.Stop()
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+	defer cancel()
+
+	var resBody echoResBody
+	_, err := client.Call(
+		ctx,
+		yarpc.NewReqMeta().Procedure("echo"),
+		&echoReqBody{},
+		&resBody,
+	)
+	assert.NoError(t, err)
+
+	AssertDepth1Spans(t, tracer)
+}
+
+func TestTChannelTracer(t *testing.T) {
+	tracer := mocktracer.New()
+
+	// Establish the TChannel
+	ch, err := tchannel.NewChannel("yarpc-test", &tchannel.ChannelOptions{
+		Tracer: tracer,
+	})
+	assert.NoError(t, err)
+	hp := "127.0.0.1:4040"
+	ch.ListenAndServe(hp)
+
+	rpc := yarpc.NewDispatcher(yarpc.Config{
+		Name: "yarpc-test",
+		Inbounds: []transport.Inbound{
+			ytchannel.NewInbound(ch),
+		},
+		Outbounds: transport.Outbounds{
+			"yarpc-test": ytchannel.NewOutbound(ch, ytchannel.HostPort(hp)),
+		},
+		Tracer: tracer,
+	})
+
+	json.Register(rpc, json.Procedure("echo", echo))
+
+	client := json.New(rpc.Channel("yarpc-test"))
+
+	rpc.Start()
+	defer rpc.Stop()
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+	defer cancel()
+
+	var resBody echoResBody
+	_, err = client.Call(
+		ctx,
+		yarpc.NewReqMeta().Procedure("echo"),
+		&echoReqBody{},
+		&resBody,
+	)
+	assert.NoError(t, err)
+
+	AssertDepth1Spans(t, tracer)
+}
+
+func AssertDepth1Spans(t *testing.T, tracer *mocktracer.MockTracer) {
+	assert.Equal(t, 2, len(tracer.FinishedSpans()), "generates inbound and outband spans")
+	if len(tracer.FinishedSpans()) != 2 {
+		return
+	}
+	spans := tracer.FinishedSpans()
+	parent := spans[0]
+	child := spans[1]
+	parentctx := parent.Context().(mocktracer.MockSpanContext)
+	childctx := child.Context().(mocktracer.MockSpanContext)
+	assert.Equal(t, parentctx.TraceID, childctx.TraceID)
+	// Whether the parent and child have the same span id is an implementation
+	// detail of the tracer.
+	assert.Equal(t, "echo", parent.OperationName, "span has correct operation name")
+	assert.Equal(t, "echo", child.OperationName, "span has correct operation name")
+}
+
+func TestHttpTracerDepth2(t *testing.T) {
+	tracer := mocktracer.New()
+
+	ctx := context.Background()
+	ctx, cancel := context.WithTimeout(ctx, 100*time.Millisecond)
+	defer cancel()
+
+	span := tracer.StartSpan("test")
+	// no defer span.Finish()
+	span.SetBaggageItem("weapon", "knife")
+	ctx = opentracing.ContextWithSpan(ctx, span)
+
+	rpc := yarpc.NewDispatcher(yarpc.Config{
+		Name: "yarpc-test",
+		Inbounds: []transport.Inbound{
+			http.NewInbound(":8080"),
+		},
+		Outbounds: transport.Outbounds{
+			"yarpc-test": http.NewOutbound("http://127.0.0.1:8080"),
+		},
+		Tracer: tracer,
+	})
+
+	client := json.New(rpc.Channel("yarpc-test"))
+
+	echo := func(ctx context.Context, reqMeta yarpc.ReqMeta, reqBody *echoReqBody) (*echoResBody, yarpc.ResMeta, error) {
```

Comment:
`echo` and `echoecho` are exactly the same between the transports, can we use local functions (or maybe a custom type that takes the `client`) rather than duplicating them?